packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersProvider.js

```diff
@@ -0,0 +1,82 @@
+const { SpvChain } = require('@dashevo/dash-spv');
+
+const BlockHeadersReader = require('./BlockHeadersReader');
+
+/**
+ * @typedef {BlockHeadersProviderOptions} BlockHeadersProviderOptions
+ * @property {string} [network=testnet]
+ * @property {boolean} [autoStart=false]
+ * @property {number} [maxParallelStreams=5] max parallel streams to read historical block headers
+ * @property {number} [targetBatchSize=100000] a target batch size per stream
+ */
+const defaultOptions = {
+  network: 'testnet',
+  autoStart: false,
+  maxParallelStreams: 5,
+  targetBatchSize: 100000,
+  fromBlockHeight: 1,
+};
+
+class BlockHeadersProvider {
+  /**
+   * @param {BlockHeadersProviderOptions} options
+   */
+  constructor(options = {}) {
+    this.options = {
+      ...defaultOptions,
+      ...options,
+    };
+
+    // TODO: Dash SPV does not understand 'regtest'. Fix
+    const network = this.options.network === 'regtest' ? 'devnet' : this.options.network;
+    this.spvChain = new SpvChain(network);
+  }
+
+  /**
+   * @param {CoreMethodsFacade} coreMethods
+   */
+  setCoreMethods(coreMethods) {
+    this.coreMethods = coreMethods;
+  }
+
+  async start() {
+    if (!this.coreMethods) {
+      throw new Error('Core methods have not been provided. Please use "setCoreMethods"');
+    }
+
+    const { chain: { blocksCount: bestBlockHeight } } = await this.coreMethods.getStatus();
+    const blockHeadersReader = new BlockHeadersReader(
+      {
+        coreMethods: this.coreMethods,
+        maxParallelStreams: this.options.maxParallelStreams,
+        targetBatchSize: this.options.targetBatchSize,
+      },
+    );
+
+    blockHeadersReader.on(BlockHeadersReader.EVENTS.BLOCK_HEADERS, (headers) => {
+      this.spvChain.addHeaders(headers.map((header) => Buffer.from(header)));
```

Comment:
We need to handle verification errors here and request headers from other nodes if needed

Reply:
Done. BlockHeadersReader catches this error and destroys the stream. Retry attempt goes afterwards

Reply:
Reworked with the explicit catch of the error from dash-spv